SYS_ATL/new_eff.py

```diff
@@ -1262,6 +1273,290 @@ def bds(x,hi):
+        pt_e    = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name) 
+        is_written      = is_elem(pt, WrG)
+        is_unchanged    = G(AEq(pt_e, stmtsG(pt_e)))
+        is_read_post    = is_elem(pt, RdGp)
+        is_overwritten  = is_elem(pt, WrGp)
+
+        # if the value of the global might be read,
+        # then it must not have been changed.
+        safe_write  = AImplies( AMay(is_read_post), ADef(is_unchanged) )
+        if not slv.verify(safe_write):
+            # for s in stmts:
+            #     print(s)
+            # print('####')
+            # print(safe_write)
+            # print('***')
+            # print(G)
+            # print('***')
+            # print(stmtsG)
+            slv.pop()
+            raise SchedulingError(
+                f"Cannot change configuration value of {pt.name} "
+                f"at {stmts[0].srcinfo}; the new (and different) "
+                f"values might be read later in this procedure")
+        # the write is invisible if its definitely unchanged or definitely
+        # overwritten
+        invisible   = ADef( AOr( is_unchanged, is_overwritten ) )
+        if not slv.verify(invisible):
+            cfg_mod_visible.add( pt.name )
+
+    slv.pop()
+    return cfg_mod_visible
+
+
+# This equivalence check assumes that we can
+# externally verify that substituting stmts with something else
+# is equivalent modulo the keys in `cfg_mod`, so
+# the only thing we want to check is whether that can be
+# extended, and if so, modulo what set of output globals?
+def Check_ExtendEqv(proc, stmts0, stmts1, cfg_mod):
+    assert len(stmts0) > 0
+    assert len(stmts1) > 0
+    ctxt = ContextExtraction(proc, stmts0)
+
+    p       = ctxt.get_control_predicate()
+    G       = ctxt.get_pre_globenv()
+    ap      = ctxt.get_posteffs()
+    #a       = G(stmts_effs(stmts))
+    sG0     = globenv(stmts0)
+    sG1     = globenv(stmts1)
+
+    slv     = SMTSolver(verbose=False)
+    slv.push()
+    #slv.assume(AMay(p))
+
+    # extract effects
+    #WrG, Mod    = getsets([ES.WRITE_G, ES.MODIFY], a)
+    WrGp, RdGp  = getsets([ES.WRITE_G, ES.READ_G], ap)
+
+    # check that none of the configuration variables which might have
+    # changed are being observed.
+    def make_point(key):
+        cfg, fld    = reverse_config_lookup(key)
+        typ         = cfg.lookup(fld)[1]
+        return APoint(key, [], typ)
+    cfg_mod_pts = [ make_point(key) for key in cfg_mod ]
+    cfg_mod_visible = set()
+    for pt in cfg_mod_pts:
+        pt_e = ABool(pt.name) if pt.typ == T.bool else AInt(pt.name)
+        is_unchanged    = AImplies(p, G(AEq(sG0(pt_e), sG1(pt_e))))
+        is_read_post    = is_elem(pt, RdGp)
+        is_overwritten  = is_elem(pt, WrGp)
+
+        safe_write  = AImplies( AMay(is_read_post), ADef(is_unchanged) )
+        if not slv.verify(safe_write):
+            slv.pop()
+            raise SchedulingError(
+                f"Cannot rewrite at {stmts[0].srcinfo} because the "
```

Comment:
There is no variable in scope named `stmts`. This should be fixed and a test should be added that covers this code.

Reply:
Thanks; did not add test.  Sorry